File: frontend/src/api/dataHealth.ts
import http from './index'

export interface DataHealthIssue {
  check: string
  severity: 'error' | 'warning' | 'hint'
  entity_type: string
  entity_id: number
  entity_uid?: string | null
  entity_name: string
  title: string
  detail: string
  suggestion: string
  deep_link?: string | null
  /** 「标注无问题」用：key 定位问题，signature 描述当前事实（数据变则豁免失效） */
  issue_key: string
  signature: string
  /** 曾被标注无问题、但数据已变化 → 重新提示 */
  reopened: boolean
}

export interface DataHealthReport {
  score: number
  counts: { error: number; warning: number; hint: number }
  total_issues: number
  ignored_count: number
  entity_stats: {
    groups: number
    artists: number
    albums: number
    songs: number
    companies: number
    music_videos: number
  }
  checks: { key: string; name: string; severity: string; count: number }[]
  issues: DataHealthIssue[]
  ignored_issues: DataHealthIssue[]
}

export interface HealthIgnoreItem {
  key: string
  signature: string
  check: string
  entity_type: string
  entity_id: number
  title: string
  note: string
  created_at: string
}

export interface PurgeDanglingResult {
  memberships_deleted: number
  mv_subjects_cleared: number
  subunit_parents_cleared: number
  artist_company_relations_deleted: number
  group_company_relations_deleted: number
}

export const dataHealthApi = {
  getReport() {
    return http.get<DataHealthReport>('/data-health').then((r) => r.data)
  },
  purgeDangling() {
    return http.post<PurgeDanglingResult>('/data-health/purge-dangling').then((r) => r.data)
  },
  /** 把某条问题标注为「无问题」：不再提示、不计分；数据变化后会自动重新提示 */
  ignoreIssue(issue: DataHealthIssue) {
    return http
      .post<{ items: HealthIgnoreItem[] }>('/data-health/ignores', {
        key: issue.issue_key,
        signature: issue.signature,
        check: issue.check,
        entity_type: issue.entity_type,
        entity_id: issue.entity_id,
        title: issue.title,
      })
      .then((r) => r.data)
  },
  /** 恢复某条问题的提示 */
  restoreIssue(key: string) {
    return http
      .post<{ items: HealthIgnoreItem[] }>('/data-health/ignores/remove', { key })
      .then((r) => r.data)
  },
  /** 恢复全部提示 */
  clearIgnores() {
    return http
      .post<{ items: HealthIgnoreItem[] }>('/data-health/ignores/clear')
      .then((r) => r.data)
  },
}

File: backend/app/schemas/data_health.py
"""数据体检报告 Schema（只读）。"""

from __future__ import annotations

from typing import List, Optional

from pydantic import BaseModel


class HealthIssue(BaseModel):
    check: str
    severity: str  # error / warning / hint
    entity_type: str
    entity_id: int
    entity_uid: Optional[str] = None
    entity_name: str = ""
    title: str
    detail: str = ""
    suggestion: str = ""
    deep_link: Optional[str] = None
    # 「标注无问题」用：key 定位问题条目，signature 描述当前事实（数据变则豁免失效）
    issue_key: str = ""
    signature: str = ""
    reopened: bool = False


class HealthCheckSummary(BaseModel):
    key: str
    name: str
    severity: str
    count: int


class HealthCounts(BaseModel):
    error: int
    warning: int
    hint: int


class HealthEntityStats(BaseModel):
    groups: int
    artists: int
    albums: int
    songs: int
    companies: int
    music_videos: int


class DataHealthReport(BaseModel):
    score: int
    counts: HealthCounts
    total_issues: int
    ignored_count: int = 0
    entity_stats: HealthEntityStats
    checks: List[HealthCheckSummary]
    issues: List[HealthIssue]
    # 已「标注无问题」的条目（不计分、不计问题数）
    ignored_issues: List[HealthIssue] = []


class HealthIgnoreItem(BaseModel):
    """一条「标注无问题」豁免。"""

    key: str
    signature: str = ""
    check: str = ""
    entity_type: str = ""
    entity_id: int = 0
    title: str = ""
    note: str = ""
    created_at: str = ""


class HealthIgnoreList(BaseModel):
    items: List[HealthIgnoreItem]


class HealthIgnoreCreate(BaseModel):
    key: str
    signature: str
    check: str = ""
    entity_type: str = ""
    entity_id: int = 0
    title: str = ""
    note: str = ""


class HealthIgnoreRemove(BaseModel):
    key: str
